File: .tmp_sc4013/java_22/3972_valid_subarrays_with_matching_sum_digits_ii.java
// CONFIG class=Solution method=countValidSubarrays types=None
// LeetCode 3972 - Valid Subarrays With Matching Sum Digits II
// https://leetcode.com/problems/valid-subarrays-with-matching-sum-digits-ii/

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public long countValidSubarrays(int[] nums, int x) {
        List<Long>[] byRemainder = new ArrayList[10];
        for (int i = 0; i < 10; i++) byRemainder[i] = new ArrayList<>();
        byRemainder[0].add(0L);
        long prefix = 0, answer = 0;
        for (int value : nums) {
            prefix += value;
            int required = (int) ((prefix - x) % 10 + 10) % 10;
            List<Long> values = byRemainder[required];
            for (long power = 1; (long) x * power <= prefix; power *= 10) {
                long low = (long) x * power;
                long high = (long) (x + 1) * power - 1;
                long minPrefix = prefix - high, maxPrefix = prefix - low;
                int left = lowerBound(values, minPrefix);
                int right = upperBound(values, maxPrefix);
                answer += right - left;
                if (power > prefix / 10) break;
            }
            byRemainder[(int) (prefix % 10)].add(prefix);
        }
        return answer;
    }

    private int lowerBound(List<Long> a, long x) {
        int lo = 0, hi = a.size();
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (a.get(mid) < x) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    private int upperBound(List<Long> a, long x) {
        int lo = 0, hi = a.size();
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (a.get(mid) <= x) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
}

File: .tmp_rs4013/rust_check22/3972_valid_subarrays_with_matching_sum_digits_ii.rs
struct Solution;
// LeetCode 3972 - Valid Subarrays With Matching Sum Digits II
// https://leetcode.com/problems/valid-subarrays-with-matching-sum-digits-ii/

impl Solution {
    pub fn count_valid_subarrays(nums: Vec<i32>, x: i32) -> i64 {
        let mut by_remainder = vec![Vec::new(); 10];
        by_remainder[0].push(0i64);
        let mut prefix = 0i64;
        let mut answer = 0i64;
        for &value in &nums {
            prefix += value as i64;
            let required = ((prefix - x as i64) % 10 + 10) % 10;
            let values = &by_remainder[required as usize];
            let mut power = 1i64;
            while x as i64 * power <= prefix {
                let low = x as i64 * power;
                let high = (x as i64 + 1) * power - 1;
                let min_prefix = prefix - high;
                let max_prefix = prefix - low;
                let left = values.partition_point(|&v| v < min_prefix);
                let right = values.partition_point(|&v| v <= max_prefix);
                answer += (right - left) as i64;
                if power > prefix / 10 {
                    break;
                }
                power *= 10;
            }
            by_remainder[(prefix % 10) as usize].push(prefix);
        }
        answer
    }
}

fn main() {}
